### ISSUE TITLE
Property can contain Doctrine\Common\Collections\Collection but database expects Doctrine\Common\Collections\Collection

### ISSUE THREAD

**Issue by andreybolonin** [phpstan/phpstan-doctrine#94]
<img width="1097" alt="Screen Shot 2019-12-09 at 6 13 51 PM" src="https://user-images.githubusercontent.com/2576509/70452399-bd8db080-1aaf-11ea-9a50-51eb9d6661ca.png">

Any ideas how to fix this? :)
I updated phpstan to 0.12.2 and use level=2

**Comment by ondrejmirtes** [comment]
Hi, this is because you should have more precise `@var` above your property, something like `@var Collection<Lesson>`. Otherwise PHPStan is not able to check if you're not assigning a wrong collection to the property.

**Comment by hxv** [comment]
I have exactly the same problem.
If I annotate var as `Collection<Type>` then I get an error:
```
Generic type Doctrine\Common\Collections\Collection<App\Entity\Type> in PHPDoc tag @var for property App\Entity\SomeName::$property does not specify all template types of class Doctrine\Common\Collections\Collection: TKey, T
```

Changing annotation to `Collection<mixed, Type>` results in error:
```
Property App\Entity\SomeName::$property type mapping mismatch: database can contain Doctrine\Common\Collections\Collection&iterable<App\Entity\Type> but property expects Doctrine\Common\Collections\Collection<mixed, App\Entity\Type>.
```

It looks to me that problem starts when column type is set in [EntityRelationRule.php](EntityRelationRule.php#L94):
```php
$columnType = TypeCombinator::intersect(
	new ObjectType('Doctrine\Common\Collections\Collection'),
	new IterableType(new MixedType(), new ObjectType($associationMapping['targetEntity']))
);
```

Shouldn't this be changed to generic type?
```php
$columnType = new GenericObjectType('Doctrine\Common\Collections\Collection', [
	new MixedType(),
	new ObjectType($associationMapping['targetEntity']),
]);
```

**Comment by ondrejmirtes** [comment]
These are two separate issues. Will look into yours @hxv.

**Comment by hxv** [comment]
My issue results from proposed solution - if `Collection` type is specified then there is error informing about type mismatch.
As far as I can tell there is no way to specify type to satisfy both PHPStan "core" and Doctrine extension.

But I can create new issue if that is necessary.

**Comment by ondrejmirtes** [comment]
Fixed: https://github.com/phpstan/phpstan-doctrine/commit/8151e1a258897809dd73806f2070ea0def787c39

The issue with "does not specify all template types" will remain because with generic stubs, `Collection<Foo>` is not valid.

**Comment by ondrejmirtes** [comment]
Just released as 0.12.2, I encourage you to try it out.

**Comment by andreybolonin** [comment]
@ondrejmirtes thx!
